### PR TITLE
Harden UnslothVisionDataCollator audio extraction and add tests

### PR DESCRIPTION
**Pull request by danielhanchen**
Follow-ups to #723 covering the remaining audio input forms that either failed silently or crashed with confusing errors. All are small additions to the extraction path in `vision_utils.py`, plus a hermetic CPU test suite for the audio collator which previously had no coverage.

**Fixes**

1. `{"type": "audio", "url": ...}` and `{"type": "audio", "path": ...}` content parts were silently dropped: the chat template still emitted the `<|audio|>` placeholder but no clip reached the processor, so the example trained as text only (same failure class as unslothai/unsloth#6004). The feature extractor already accepts path and URL strings, so these keys are now resolved. An audio part with no usable payload now raises instead of silently training text only.
2. HF Audio dicts with a `sampling_rate` different from the feature extractor's were processed at the wrong speed with no warning (a 1s clip at 44.1kHz became 2.75s of 16kHz audio). This now raises with a pointer to `dataset.cast_column("audio", Audio(sampling_rate=16000))`. Resampling in the collator was deliberately avoided to keep dependencies unchanged.
3. `torch.Tensor` clips (what `torchaudio.load` returns) crashed inside the feature extractor with a cryptic broadcast error. They are now converted to numpy. Stereo 2-D arrays now raise a clear "convert to mono" error instead of the same cryptic failure.

**Tests**

`tests/test_vision_collator_audio.py`: 17 hermetic CPU tests with stub processors covering `extract_audio_info` (inline arrays, HF dicts, url/path, payload-less raise, rate validation), `_extract_audio_for_example` (top-level dict/ndarray/flat list/list-of-clips columns, None fallback, tensor conversion, mono guard) and `_truncate_sequence_tensors` (per-token-key allowlist so audio feature tensors are never sliced on dimension collisions, left-padding aware slicing, audio-span truncation guard).

**Verification**

- All 17 new tests pass.
- Verified against the real Gemma 4 processor on transformers 5.10.2: url/path parts produce correct `input_features`, the three new errors fire with clear messages, torch tensors work, and the plain ndarray happy path is unchanged (25 placeholder tokens, 99 frames for a 1s clip).
- Re-ran the full collator comparison harness against main: text only, image only, image truncation and prompt-completion batches remain bit identical, and the truncation collision cases stay fixed.

Decisions worth noting from the review of #723: audio boundary tokens (`<|audio>`, `<audio|>`) intentionally stay trainable in labels, matching the image policy where `<|image>` and `<image|>` are trainable while soft tokens are masked. The stale prompt-length `mm_token_type_ids` in the prompt-completion path is pre-existing, only consumed by the larger Gemma 4 variants with bidirectional vision attention, and is left for a separate fix.

### CODE REVIEW

**Comment by danielhanchen** [comment]
Addressed the two review findings in the latest commit:

1. Mono `torchaudio.load` tensors: `torchaudio.load` returns `[channels, frames]`, so a mono file is `(1, N)`. `_normalize_audio_clips` now squeezes single-channel 2-D clips to 1-D and only raises for genuinely multi-channel audio. Verified end to end with the real Gemma 4 processor: `(1, 16000)` tensor produces `input_features (1, 99, 128)`.
2. Top-level `path`/`url` dict payloads: dict resolution is now shared between the inline and top-level paths via `_resolve_audio_dict`, so `{"audio": {"bytes": None, "path": ...}}` (the `Audio(decode=False)` shape) and list entries like `[{"path": ...}]` resolve to the path string the feature extractor accepts. Payload-less dicts now raise the same clear error as inline parts instead of silently dropping.

Added 5 regression tests for these cases (22 total, all passing), and re-verified the unchanged happy path with the real processor.

**Comment by danielhanchen** [comment]
Second review round addressed:

1. Stereo serialized as nested Python lists (for example after `.tolist()` or JSON round-trips) now hits the same mono guard as tensors and ndarrays: list and tuple clips are converted with `np.asarray` inside `_normalize_audio_clips`, so `[[left], [right]]` raises the clear mono error and a nested mono `[[samples]]` is squeezed to 1D. Note the top-level `audio` column keeps its documented list-of-clips meaning, so `{"audio": [[c1], [c2]]}` is still two clips; the placeholder count check covers genuine mismatches there.
2. A top-level list of path or URL strings (`{"audio": ["a.wav", "b.wav"]}`) was misclassified as one flat clip of "samples". Strings are now classified as clips, so each path resolves separately. Verified with the real Gemma 4 processor: two path clips with two placeholders produce `input_features (2, 99, 128)`.

Added 5 more regression tests (26 total, all passing) and re-verified the unchanged happy path.

File: unsloth_zoo/vision_utils.py [hunk 1]
```diff
@@ -932,28 +957,47 @@ def _extract_images_videos_for_example(self, example, messages):
...
+            # Raw ndarray or other single-clip value
+            clips = [audio_val]
+        return self._normalize_audio_clips(clips)
+
+    def _normalize_audio_clips(self, clips):
+        normalized = []
+        for clip in clips:
+            if torch.is_tensor(clip):
+                clip = clip.detach().cpu().numpy()
+            if getattr(clip, "ndim", 1) > 1:
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/vision_utils.py:1026]: Already handled as of 2919b925: list and tuple clips are converted with np.asarray before the ndim check, so nested-list stereo hits the mono guard and nested mono is squeezed to 1D. 050782aa additionally rejects string lists inside a single clip with a clear error instead of letting a string ndarray reach the feature extractor. Covered by tests in tests/test_vision_collator_audio.py.
